[receiver/mongodbatlas] Add alerts receiver

BinaryFissionGames opened the pull request:
**Description:**
* Adds support for receiving mongodbatlas alerts via webhooks

**Link to tracking Issue:** #10488

**Testing:**
* Unit tests added for new alerts receiver
* Integration test that runs the new alerts receiver and makes a client request using sample payloads.

**Documentation:**
* Added info to README for alerts configuration

BinaryFissionGames committed:
add alerts receiver

BinaryFissionGames committed:
docs, change defaults to make sense

BinaryFissionGames committed:
add config test for validate

BinaryFissionGames committed:
remove write logs

BinaryFissionGames committed:
remove observed timestamp check

BinaryFissionGames committed:
no need to mock now

BinaryFissionGames committed:
fix linting issues

BinaryFissionGames committed:
add integration build constraint to integration test

BinaryFissionGames committed:
use factory methods for integration test

BinaryFissionGames committed:
changelog

BinaryFissionGames committed:
fix some weird wording in the README

cpheps commented on receiver/mongodbatlasreceiver/alerts.go:
```diff
@@ -0,0 +1,316 @@
+		return
+	}
+
+	payloadSigHeader := req.Header.Get(signatureHeaderName)
```
nit: Might be worth checking the headers before trying to read in the payload. Fail fast mentality.

cpheps commented on receiver/mongodbatlasreceiver/alerts.go:
```diff
@@ -0,0 +1,316 @@
+	}
+
+	payload := make([]byte, req.ContentLength)
+	_, err := io.ReadFull(req.Body, payload)
```
The server should close the request `Body`. `ReadFull` takes in an `io.Reader` so it won't close for you.

BinaryFissionGames replied:
Right, but I don't think we have to close the Body in the handler, the server code handles that for us, see the comment on the Body field here: https://pkg.go.dev/net/http#Request
```
	// For server requests, the Request Body is always non-nil
	// but will return EOF immediately when no body is present.
	// The Server will close the request body. The ServeHTTP
	// Handler does not need to.
```

cpheps commented on receiver/mongodbatlasreceiver/alerts.go:
```diff
@@ -0,0 +1,316 @@
+	return pcommon.Timestamp(0)
+}
+
+func severityFromAlert(a model.Alert) plog.SeverityNumber {
```
Is there a doc link we can post as a comment in here just for a reference for any one who looks at this?

cpheps commented on receiver/mongodbatlasreceiver/internal/model/alerts.go:
```diff
@@ -0,0 +1,41 @@
+
+package model
+
+type Alert struct {
```
Need comment on this struct.  Also, can we include documentation on this structure for future devs?

cpheps commented on receiver/mongodbatlasreceiver/testdata/alerts/golden/metric-threshold-closed.json:
```diff
@@ -0,0 +1 @@
+{"resourceLogs":[{"resource":{"attributes":[{"key":"mongodbatlas.group.id","value":{"stringValue":"REDACTED-group"}},{"key":"mongodbatlas.alert.config.id","value":{"stringValue":"REDACTED-alert"}},{"key":"mongodbatlas.cluster.name","value":{"stringValue":"Cluster0"}},{"key":"mongodbatlas.replica_set.name","value":{"stringValue":"REDACTED-shard-0"}}]},"scopeLogs":[{"scope":{},"logRecords":[{"timeU ...
```
nit: newline

djaglowski replied:
Please pretty print this.

cpheps commented on receiver/mongodbatlasreceiver/testdata/alerts/golden/new-primary.json:
```diff
@@ -0,0 +1 @@
+{"resourceLogs":[{"resource":{"attributes":[{"key":"mongodbatlas.group.id","value":{"stringValue":"REDACTED-group"}},{"key":"mongodbatlas.alert.config.id","value":{"stringValue":"REDACTED-alert"}},{"key":"mongodbatlas.cluster.name","value":{"stringValue":"Cluster0"}},{"key":"mongodbatlas.replica_set.name","value":{"stringValue":"REDACTED-shard-0"}}]},"scopeLogs":[{"scope":{},"logRecords":[{"timeU ...
```
nit: newline

BinaryFissionGames committed:
PR feedback

BinaryFissionGames commented:
cc: @zenmoto @djaglowski

BinaryFissionGames committed:
make goporto

djaglowski commented on receiver/mongodbatlasreceiver/alerts.go:
```diff
@@ -0,0 +1,316 @@
+
+	attrs := logRecord.Attributes()
+	// These attributes are always present
+	attrs.InsertString("event.name", alert.EventType)
```
This aligns well with [otep 202](https://github.com/open-telemetry/oteps/pull/202), which is not yet accepted but seems like to be soon.

Another thing that's called out in that otep is that `event.name` is only meaningful within an `event.domain`, so I think we should probably add that here as well. (Can be hardcoded as `mongoatlas`, or something similar.)

BinaryFissionGames replied:
I'm a little confused on what "Scope Level" means in context of pdata, but what I got from there is that it's a log-record level attribute, so I've added it there.

djaglowski replied:
Yeah, "scope" is a newer term, replacing "instrumentation library". Additionally, there are now going to be attributes at the scope level.

So you have resource -> scope -> logrecord, each of which has its own attributes. In the context of logs, scope will commonly refer to an instance of a logger.

djaglowski commented on receiver/mongodbatlasreceiver/alerts_integration_test.go:
```diff
@@ -0,0 +1,253 @@
+				sink,
+			)
+
+			require.NoError(t, err)
+
+			require.NoError(t, err)
```
```suggestion
			require.NoError(t, err)
```

djaglowski commented on receiver/mongodbatlasreceiver/alerts_integration_test.go:
```diff
@@ -0,0 +1,253 @@
+			require.NoError(t, err)
+
+			defer func() {
+				assert.NoError(t, recv.Shutdown(context.Background()))
```
Why `assert` vs `require` here?

BinaryFissionGames replied:
I was incorrectly under the impression that defer blocks had some weird interaction with t.FailNow(), but it doesn't seem to be a problem. I think I was confusing it with goroutines.

djaglowski commented on receiver/mongodbatlasreceiver/alerts_integration_test.go:
```diff
@@ -0,0 +1,253 @@
+
+			require.Eventually(t, func() bool {
+				return sink.LogRecordCount() > 0
+			}, 10*time.Second, 10*time.Millisecond)
```
nit: Unless really necessary, can the timeout be shorter? 10s isn't a lot, but if probably 5-8s longer than necessary for the vast majority of failure cases.

BinaryFissionGames replied:
bumped this down to 2 seconds.

djaglowski commented on receiver/mongodbatlasreceiver/alerts_integration_test.go:
```diff
@@ -0,0 +1,253 @@
+			expectedLogs, err := readLogs(filepath.Join("testdata", "alerts", "golden", payloadName))
+			require.NoError(t, err)
+
+			require.NoError(t, compareLogs(*expectedLogs, logs))
```
Nice. This would be a great addition to `scrapertest` (likely with a rename to that package).

BinaryFissionGames replied:
Opened #10896 to track adding these functions

djaglowski replied:
Awesome, thank you.

djaglowski commented on receiver/mongodbatlasreceiver/testdata/alerts/cert/cert.pem:
```diff
@@ -0,0 +1,31 @@
+-----BEGIN CERTIFICATE-----
```
Please integrate with the [`certs` make target](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/Makefile#L335).

BinaryFissionGames commented on receiver/mongodbatlasreceiver/README.md:
```diff
@@ -1,33 +1,50 @@
   - `initial_interval` (default 5s)
   - `max_interval` (default 30s)
   - `max_elapsed_time` (default 5m)
+- `alerts`
```
I did propose a couple different configs in the tracking issue (#10488) - it does feel weird that the configuration is asymmetric. Of course, backwards compatibility is the main concern, as you noted.

BinaryFissionGames committed:
More PR feedback

djaglowski commented on Makefile:
```diff
@@ -330,7 +330,8 @@ endef
              receiver/signalfxreceiver/testdata \
-             receiver/splunkhecreceiver/testdata
+             receiver/splunkhecreceiver/testdata \
+			 receiver/mongodbatlasreceiver/testdata/alerts/certs
```
Please fix indentation

djaglowski reviewed (commented):
Please also run `gotidy` to resolve the failing check:
```
go.mod/go.sum deps changes detected, please run "make gotidy" and commit the changes in this PR.
```

BinaryFissionGames committed:
fix tab/space mixing

djaglowski commented on receiver/mongodbatlasreceiver/README.md:
```diff
@@ -1,33 +1,50 @@
+as well as alerts via a configured [webhook](https://www.mongodb.com/docs/atlas/tutorial/third-party-service-integrations/).
 
-Supported pipeline types: metrics
+Supported pipeline types: metrics, logs
```
There's a new component readme header format that you'll need to use instead. See #10006

BinaryFissionGames committed:
rework to use GetAvailalableLocalAddress

BinaryFissionGames committed:
Add status header

BinaryFissionGames commented:
I added in the status header; Marked as beta since that's where it seems to be.

I also moved over from GetAvailablePort to GetAvailableAddress in anticipation of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10871

@zenmoto - mind giving this a review?

djaglowski reviewed (approved):
LGTM.

@zenmoto do you want to give this a look as codeowner of the component?

djaglowski commented:
@zenmoto, I'll merge this later today. Please let me know if you have any concerns.
